server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/VariableWidthHistogramAggregator.java

```diff
@@ -0,0 +1,278 @@
+                if(values.advanceExact(doc)){
+                    final int valuesCount = values.docValueCount();
+                    double prevVal = Double.NEGATIVE_INFINITY;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        double val = values.nextValue();
+                        assert val >= prevVal;
+                        if (val == prevVal){
+                            continue;
+                        }
+                        DocInfo docInfo = new DocInfo(doc, val, sub);
+                        collectValue(docInfo);
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * <code>docInfo</code> is stored in memory until there are <code>cacheLimit</code> documents.
+     * Then, these documents are clustered into <code>3/4 * shardSize</code> buckets.
+     * Subsequent documents will either be be collected into the nearest existing bucket, or into a new bucket if
+     * their value is distant from all of the existing buckets.
+     */
+    private void collectValue(DocInfo docInfo) throws IOException{
+        if(cachedDocs.size() < cacheLimit) {
+            cachedDocs.add(docInfo);
+            if(cachedDocs.size() == cacheLimit) {
+                // The cache is full -> Collect the cached set of documents into buckets
+                bucketAndSortCachedDocs();
+
+                // Calculate the average distance between buckets.
+                // Subsequent documents that are collected will be compared with this value to determine if they should
+                // be collected into an existing bucket or into a new bucket.
+                avgBucketDistance = 0;
+
+                // This calculation assumes that buckets are sorted in increasing order
+                for(int i = 0; i < buckets.size() - 1; i++){
+                    avgBucketDistance += buckets.get(i + 1).centroid - buckets.get(i).centroid;
+                }
+                avgBucketDistance /= buckets.size() - 1;
+            }
+        } else {
+            int bucketOrd = getNearestBucket(docInfo.val);
+            double distance = Math.abs(buckets.get(bucketOrd).centroid - docInfo.val);
+            if(distance > (2 * avgBucketDistance) && buckets.size() < shardSize) {
+                // Make a new bucket since the document is distant from all existing buckets
+
+                // Ideally, a new bucket should be created whenever a document is distant, even if we have more
+                // than shardSize buckets. Then, we should merge buckets after collection so that the aggregator
+                // produces at most shardSize buckets.
+                // However, merging nearby buckets would require continually tracking nearby buckets.
+                // This would get complicated. For now, setting a high shard size will resolve this issue.
+                // TODO: Add this merging logic which will improve the bucketing performance
+
+                buckets.add(
```

Comment:
Just as an FYI, we try to defer actual bucket object creation until the very end (in `buildAggregation()`), since they eat up a fair amount of memory compared to the data being aggregated (in this case, the values and the centroids).  You'll see that the aggregators only work with the raw data primitives in hashmaps and arrays until all the data is collected, then the final set of buckets are generated.